### PR TITLE
Select and pull a devfile using Alizer

### LINKED ISSUE #5408: `odo init` - add ability to create component with existing codebase
/kind user-story

## User Story
- As an odo user
- I want to easily add devfile.yml to my existing project
- So that can quickly start benefiting from Devfile ecosystem and other odo features depending on devfile (`odo deploy` `odo dev`)

## Acceptance Criteria 
- [x] The command should always work with the current directory
- [x] If the command is executed in the directory with the existing `devfile.yaml` file it should error out.
- [x] The command should have an **interactive mode** that is by default executed when the user runs the command without any flags. 
  - [x] The first line of the output should explain to odo user if odo detected the source code and will try to autodetect the project type, or if it works in a mode where it will try to bootstrap new project in an empty directory (details in examples)
  - [x] If the command is executed in the directory with existing source code it should use [Alizer](https://github.com/redhat-developer/alizer) to automatically detect correct devfile.yaml from all devfile registries configured in odo. As implemented in #5422
    - [x] After the successful detection it will show  Language and Project Type information to users and ask for confirmation. If user answers that the information is not correct, odo should ask "Select language" and "Select project type" questions.
    - [x] Users should be able to interactively add or remove ports and environment variables that are configured in devfile. This is only in the case the command is executed in the directory with existing code. as implemented in #5423 
  - [x] The last question should ask for the component name. The default value should match the current directory name (be aware that component name has a character limit and it can be only  alphanumer characters and `-`, more info in devfile spec.) user entered value should be saved as `metadata.name` in the `devile.yaml`.
- [x] The command should also have non-interactive mode that is active when flags are used with the following flags.
  - [x] `--devfile` `string` Name of the Devfile from the Devfile registry (required if --devfile-path is not defined)
  - [x] `--devfile-path` `string` Path to a devfile. It can be local filesystem path or http(s). This is alternative to using devfile from Devfile registry.
  - [x] `--name` `string` Name of the new component. If not specified, the name will be the name of the current directory.
  - [x] `--registry` `string` name of the devfile registry (as configured in odo preference registry). It can be used in combination with --devfile, but not with --devfile-path
  - [x] `--starter` `string` Name of the devfile starter project to populate the current directory with.
- [x] The command should inform the user about how it is progressing (downloading etc...) (see example for details)
- [x] The last message should help the user with next steps and (see example for details)

### when executed in  directory with existing source code

```
$ odo init

Current directory already contains source code.
Odo will try to autodetect language and project type to select best suited Devfile for your project.
Detected java using spring
? Is this correct? Yes
Current component configuration:
Container "container2":
  Opened ports:
  Environment variables:
   - FOO = bar
Container "container1":
  Opened ports:
   - 8080
   - 8084
  Environment variables:
   - FOO = bar
   - FOO1 = bar1
? Select container for which you want to change configuration? NONE - configuration is correct

? Enter component name: myjava
⠸ Downloading "java-springboot". 
⠏ Downloading "java-springboot". DONE
Your new component "myjava" is ready in the current directory.
To start editing your project, use “odo dev” and open this folder in your favorite IDE.
Changes will be directly reflected on the cluster.
To deploy your application to your cluster use “odo deploy”.

```

### when executed in empty directory

```
$ odo init

Current directory is empty. You can create a new component using a starter project to easily start new project.

? Select language: java
? Select project type: Maven Java
? Which starter project do you wan to use? starter1
? Enter component name: myjava
⠏ Downloading "java-maven". DONE
⠏ Downloading starter project "starter1" ... DONE
Your new component "myjava" is ready in the current directory.
To start editing your project, use “odo dev” and open this folder in your favorite IDE.
Changes will be directly reflected on the cluster.
To deploy your application to your cluster use “odo deploy”.

```

### non-interactive
```
$ odo init --name myjava --devfile java-maven

⠏ Downloading "java-maven". DONE

Your new component "myjava" is ready in the current directory.
To start editing your project, use “odo dev” and open this folder in your favorite IDE.
Changes will be directly reflected on the cluster.
To deploy your application to your cluster use “odo deploy”.

$ odo init --name myjava --devfile java-maven --registry my-company-registry --starter starter1

⠏ Downloading "java-maven". DONE
⠏ Downloading starter project "starter1" ... DONE

Your new component "myjava" is ready in the current directory.
To start editing your project, use “odo dev” and open this folder in your favorite IDE.
Changes will be directly reflected on the cluster.
To deploy your application to your cluster use “odo deploy”.



$ odo init --name myjava --devfile-path https://example.com/devfile.yml

⠏ Downloading "https://example.com/devfile.yml". DONE

Your new component "myjava" is ready in the current directory.
To start editing your project, use “odo dev” and open this folder in your favorite IDE.
Changes will be directly reflected on the cluster.
To deploy your application to your cluster use “odo deploy”.


$ odo init --name myjava --devfile /path/to/devfile.yml

⠏ Copying "/path/to/devfile.yml" to current directory. DONE

Your new component "myjava" is ready in the current directory.
To start editing your project, use “odo dev” and open this folder in your favorite IDE.
Changes will be directly reflected on the cluster.
To deploy your application to your cluster use “odo deploy”.
```

## Links 
- Related Epic: #5312

### LINKED ISSUE #5422: Select and pull a devfile using Alizer
/kind feature

Create a package to select a devfile and pull it, using Alizer to auto-detect the devfile
Given:
- a directory containing source files
- a destination directory to store devfile

```
Based on the files in the current directory odo detected
Language: Java
Project type: SpringBoot

? Is this correct? Yes
```

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**

- Add an alizer module
- Adds support for Alizer in `odo init` command
- Adds support for Alizer in `odo deploy` command
- Remove `--context` flag from `odo deploy` command

**Which issue(s) this PR fixes:**

Fixes #5422 
Fixes #5408 

**PR acceptance criteria:**

- [x] Unit test 

- [x] Integration test 

- [x] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: tests/integration/devfile/cmd_devfile_init_test.go [hunk 1]
```diff
@@ -46,6 +56,19 @@ var _ = Describe("odo devfile init command tests", func() {
 		})
 	})
 
+	When("running odo init from a directory with sources", func() {
+		BeforeEach(func() {
+			helper.CopyExample(filepath.Join("source", "nodejs"), commonVar.Context)
+		})
+		It("should work without --starter flag", func() {
+			helper.Cmd("odo", "init", "--name", "aname", "--devfile", "go").ShouldPass()
```
Comment (valaparthvi) [inline tests/integration/devfile/cmd_devfile_init_test.go:64]:
```suggestion
			helper.Cmd("odo", "init", "--name", "aname", "--devfile", "nodejs").ShouldPass()
```
Or was using a Go devfile in a nodejs project intentional?

File: pkg/init/backend/alizer.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+// SelectDevfile calls thz Alizer to detect the devfile and asks for confirmation to the user
+func (o *AlizerBackend) SelectDevfile(flags map[string]string, fs filesystem.Filesystem, dir string) (location *DevfileLocation, err error) {
+	selected, registry, err := o.detectFramework(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf("Based on the files in the current directory odo detected\nLanguage: %s\nProject type: %s\n", selected.Language, selected.ProjectType)
+	fmt.Printf("The devfile %q from the registry %q will be downloaded.\n", selected.Name, registry.Name)
```
Comment (valaparthvi) [inline pkg/init/backend/alizer.go:72]: Is this final or do you plan on using something fancy from the log pkg?

File: pkg/init/backend/interface.go [hunk 3]
```diff
@@ -6,15 +6,16 @@ package backend
...
 	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
 	"github.com/devfile/library/pkg/devfile/parser"
+	"github.com/redhat-developer/odo/pkg/testingutil/filesystem"
 )
 
 // InitBackend is a specialized backend for steps of initiating a project, based on various input (either from CLI flags or interactively from user)
 type InitBackend interface {
-	// Validate returns an error if it does not validate the flags
-	Validate(flags map[string]string) error
+	// Validate returns an error if it does not validate the flags basde on the directory content
```
Comment (valaparthvi) [inline pkg/init/backend/interface.go:14]:
```suggestion
	// Validate returns an error if it does not validate the flags based on the directory content
```

**Review by valaparthvi — COMMENTED** [review]
Changes lgtm, just a few questions.

File: pkg/init/backend/alizer.go (hunk 2 above)
Comment (feloy) [reply pkg/init/backend/alizer.go:72]: I tried different methods from log pkg, but I'm not sure which one to use

**Comment by feloy** [comment]
/test v4.9-integration-e2e

infra errors

**Review by valaparthvi — APPROVED** [review]

**Pull request merged**